### PR TITLE
Leaderboard

### PR DESCRIPTION
**Pull request by jMyles**

### CODE REVIEW

File: contracts/Contribution.sol [hunk 1]
```diff
@@ -22,10 +22,13 @@ contract Contribution {
...
     bool public testnet;
-    
+
     // contributions storage
-    mapping(address => uint256[]) public contributionsByAddress;
-    address[] public contributors;
+    bool[] public contributionIsCombined;
+    uint256[] public contributionAmounts;
+    uint256[] public contributionDatetimes;
+    address[] public contributorsForEachAContribution;
```
Comment (kubic71) [inline contracts/Contribution.sol:30]: What does the "EachA" mean? Is it like "Each address"?

File: utilities/contribution_utilities.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+def calculate_leaders(contribution_metadata) -> Dict[str, int]:
+    """
+    Takes three lists (from solidity arrays) and returns a dictionary of leaders.
+    """
+    contributors = contribution_metadata[0]
+    amounts = contribution_metadata[1]
+    combined = contribution_metadata[2]
+    datetime = contribution_metadata[3]
+
+    contributors_by_address = defaultdict(list)
```
Comment (kubic71) [inline utilities/contribution_utilities.py:14]: Should this be named maybe `contributions_by_address`?

File: contracts/Contribution.sol [hunk 3]
```diff
@@ -107,32 +110,26 @@ contract Contribution {
...
-        }
-
-        if (combine) {
-            contributionsByAddress[msg.sender][0] += msg.value;
-        } else {
-            contributionsByAddress[msg.sender].push(msg.value); // Add contribution to the mapping
-        }
+        contributionAmounts.push(msg.value);
+        contributorsForEachAContribution.push(msg.sender);
+        contributionIsCombined.push(combine);
```
Comment (kubic71) [inline contracts/Contribution.sol:121]: What is the meaning of the `combine` flag?

For example if I make the following contributions:
- contribute(0.5)
- contributeAndCombine(0.2)
- contributeAndCombine(0.4)
- contribute(0.1)
- contribute(0.25)
- contributeAndCombine(0.6)
- contribute(1)

What should my contributions be? Should it be:

### `combine` means "combine all my previous contributions into one"
Therefore I'd have 1 contribution with value `2.05` eth after the last `contributeAndCombine(0.6)` and one contribution with value `1` eth.


### `combine` flag set to true means combine with a first "default" contribution
In this case, the evolution of contributions would be like  this (`contributions` is an array of user contributions)
- contribute(0.5) - `contributions[0] = 0.5` - the default zero-index contribution gets assigned  
- contributeAndCombine(0.2) - `contributions[0] = 0.5 + 0.2 = 0.7`
- contributeAndCombine(0.4) - `contributions[0] = 0.7 + 0.4 = 1.1`
- contribute(0.1) - `contributions[0] = 1.1, contributions[1] = 0.1`
- contribute(0.25) - `contributions[0] = 1.1, contributions[1] = 0.1, contributions[2] = 0.25`
- contributeAndCombine(0.6) - `contributions[0] = 1.1 + 0.6 = 1.7, contributions[1] = 0.1, contributions[2] = 0.25` 
- contribute(1) - `contributions[0] = 1.7, contributions[1] = 0.1, contributions[2] = 0.25, contributions[3] = 1.0`

### something else?

File: utilities/contribution_utilities.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    for counter, contributor in enumerate(contributors):
+        is_combined = combined[counter]
+        amount = amounts[counter]
+
+        if is_combined:
+            if not contributors_by_address[contributor]:
+                # This ought to be an impossible situaiton - how did they dcombine with a bid that didn't exist?
```
Comment (kubic71) [inline utilities/contribution_utilities.py:22]: Right now you can call `contributeAndCombine` function of the contract as your first contribution, without previously calling `contribute`, so it's not impossible?

File: utilities/contribution_utilities.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+                contributors_by_address[contributor].append(0)
+            contributors_by_address[contributor][0] += amount
```
Comment (kubic71) [inline utilities/contribution_utilities.py:24]: This suggest we go with the "combine flag set to true means combine with a first default contribution" option, which I mentioned in my comment above:
https://github.com/JustinHolmesMusic/Revealer/pull/50/files#r1299359630

**Comment by KPrasch** [comment]
subsumed by #53

File: contracts/Contribution.sol (hunk 3 above)
Comment (jMyles) [reply contracts/Contribution.sol:121]: Seems like we have this resolved, right?  For now, it just always adds to your initial contributions.

**Pull request merged**